libfwupd/fwupd-client.c

```diff
@@ -535,7 +541,9 @@ fwupd_client_connect_async (FwupdClient *self, GCancellable *cancellable,
 	g_return_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable));
 
 	/* nothing to do */
+	g_mutex_lock (&priv->proxy_mutex);
```

Comment:
Holding the lock until the end of `fwupd_client_connect_get_proxy_cb()` means that any other pending calls to `fwupd_client_connect_async()` will block on this line. i.e. The `*_async()` call will become blocking.

I think you probably want to unlock the mutex at end of this function, and lock it again in `fwupd_client_connect_get_bus_cb()` and `fwupd_client_connect_get_proxy_cb()` when you access the shared state in either of them. You’d then have to deal with what might have happened if another thread executed their `fwupd_client_connect_get_bus_cb()`/`fwupd_client_connect_get_proxy_cb()` callback first. i.e. `priv->conn` is unexpectedly not `NULL`, or `priv->proxy` is unexpectedly not `NULL`. In both cases it should be OK to free and drop the newly-returned result from `g_bus_get_finish()` and `g_dbus_proxy_new_finish()` and move on to the next step in the process.

Reply:
That makes sense; will give it a go, thanks!

Reply:
@pwithnall something like this?